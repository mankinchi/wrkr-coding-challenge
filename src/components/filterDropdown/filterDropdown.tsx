import { ChangeEventHandler } from "react";

interface Props {
	prompt: string;
	options: {
		title: string;
		value: string;
	}[];
	value: string;
	onChange: (newValue: string) => void;
}

export const ALL_OPTION_VALUE = "all";
export type AllOptionType = typeof ALL_OPTION_VALUE;

export const FilterDropdown = (props: Props) => {
	const { prompt, options, value, onChange } = props;

	const handleOptionSelected: ChangeEventHandler<HTMLSelectElement> = (e) => {
		onChange(e.currentTarget.value);
	};

	return (
		<div>
			<span>{prompt}</span>
			<select
				value={value}
				className="focus:outline-none"
				onChange={handleOptionSelected}
			>
				{options.map(({ title, value: optionValue }) => (
					<option
						key={optionValue}
						value={optionValue}
					>
						{title}
					</option>
				))}
			</select>
		</div>
	);
};
