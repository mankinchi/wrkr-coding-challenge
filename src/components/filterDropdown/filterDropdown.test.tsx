import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { FilterDropdown } from "./filterDropdown";

describe("Filter Dropdown component", () => {
	it("should render correctly", () => {
		const onChange = jest.fn();

		const { container } = render(
			<FilterDropdown
				prompt="Filter prompt"
				options={[
					{
						title: "1",
						value: "filter 1",
					},
					{
						title: "2",
						value: "filter 2",
					},
				]}
				value="1"
				onChange={onChange}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it("should fire onChange when new value is selected", () => {
		const onChange = jest.fn();

		render(
			<FilterDropdown
				prompt="Filter prompt"
				options={[
					{
						title: "filter 1",
						value: "1",
					},
					{
						title: "filter 2",
						value: "2",
					},
				]}
				value="1"
				onChange={onChange}
			/>,
		);

		userEvent.selectOptions(screen.getByRole("combobox"), ["2"]);

		expect(screen.getByRole("combobox")).toHaveTextContent("filter 2");
		expect(onChange).toHaveBeenCalledTimes(1);
	});
});
