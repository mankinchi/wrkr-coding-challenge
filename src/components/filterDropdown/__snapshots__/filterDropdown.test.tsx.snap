// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Dropdown component should render correctly 1`] = `
<div>
  <div>
    <span>
      Filter prompt
    </span>
    <select
      class="focus:outline-none"
      data-cy="filter-select"
    >
      <option
        value="filter 1"
      >
        1
      </option>
      <option
        value="filter 2"
      >
        2
      </option>
    </select>
  </div>
</div>
`;
